staging/vhost-device-can/README.md

```diff
@@ -0,0 +1,114 @@
+# vhost-device-can - CAN emulation backend daemon
+
+## Description
+This program is a vhost-user backend that emulates a VirtIO CAN device.
+The device's binary takes two (2) parameters:  a socket, a 'can-devices' list.
+The socket is commonly used across all vhost-devices to communicate with
+the vhost-user frontend device. The 'can-devices' represents a list of
+CAN/FD devices appears in the host system. This list includes *can_input* and
+*can_output* devices which vhost-device-can will forward messages to and from
+the frontend side.
+
+This program is tested with QEMU's `vhost-user-device-pci` device.
+Examples' section below.
+
+## Synopsis
+```
+**vhost-device-can** [*OPTIONS*]
+````
+
+## Options
+
+.. program:: vhost-device-can
+
+.. option:: -h, --help
+
+  Print help.
+
+.. option:: -s, --socket-path=PATH
```

Comment:
Hello, I tried this and it seems the daemon is adding a '0' at the end of the socket path. Is this OK? Shall not the daemon to create the unix socket in exactly "PATH"?
```
 Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/vhost-device-can --socket-path /tmp/vhost-can.socket --can-devices 'vcan0:vcan0'`
thread_id: 0, socket: "/tmp/vhost-can.socket0", can_in: "vcan0", can_out: "vcan0"
[vhost-device-can/src/can.rs:165] "Start reading from {} socket!" = "Start reading from {} socket!"
[vhost-device-can/src/can.rs:165] &can_in_name = "vcan0"
```

Reply:
Hello, this normal, the user provides the general name of the socket and the device adds a number in the end of it.
That's happening because based on `socket-count=N` the device will attempt to create *N* sockets (one for each guest).                                                                               
As far as I can tell, this is the case for most of the vhost-devices (ex. vhost-device-rng).

Reply:
Yes exactly! Let's say we call vhost-device-can as:
```text
./vhost-device-can --socket-path /tmp/can.sock  --socket-count 2 --can-devices "vcan0:vcan0 vcan1:vcan2"
```
This command will start the device and create two new sockets: */tmp/can.sock0* and */tmp/can.sock1*.
From the other side we run two QEMU instances (VMs) with vhost-user-can:
```text
qemu-system-x86_64 \ 
                    ...
                    -numa node,memdev=mem1 \
                    -object memory-backend-file,id=mem1,size=2G,mem-path=/var/tmp/foo4,share=on \ 
                    -chardev socket,id=char1,path=<SOCKET_PATH> \                      
                    -device vhost-user-can-pci,chardev=char1,id=can \
                    ...
```
In the first instance of QEMU *SOCKET_PATH* would be: */tmp/can.sock0*, and will use *can0* (host interface) as sender and receiver. The second QEMU VM would have: *SOCKET_PATH = /tmp/can.sock1*, and will use *can1* as receiver and *can2* as sender.

Reply:
Cool, do you mind to add this in the `README.md` somewhere? I think that may help someone that is testing it.

Reply:
Great! I have it added in a new subsection called "Multi-Guest case"

Reply:
If socket-count is 1 the device will create `/tmp/can.sock` or `/tmp/can.sock0`?

> As far as I can tell, this is the case for most of the vhost-devices (ex. vhost-device-rng).

In `vhost-device-vsock` we require defining the name of each socket, so it is not really common, but I see why using it in this device, it should be okay.